### PR TITLE
Issue 57 data processing checks

### PR DESCRIPTION
**Pull request by widal001**
## Summary

Sets up basic CI/CD checks for scripts in `data-roadmap/` and `data-pipeline/` directories outlined in #57 

## Definition of Done

- [ ] For `data-pipeline/`
  - [x] Standup basic CI/CD for commits
  - [x] Add checks
    - [x] Linting (`black`, `flake8`, `pylint`)
    - [x] Security auditing (`safety`)
    - [x] License checking (`liccheck`)
    - [x] Unit testing (`pytest`, `pytest-coverage`)
    - [x] Integration testing (`pytest`, `pytest-coverage`)
    - [ ] Package management (`poetry`)
  - [x] Checks run against data processing scripts on every push and PR (`GitHub Actions`)
- [ ] For `data-roadmap/`
  - [x] Standup basic CI/CD for commits (`pre-commit`, `tox`)
  - [x] Add checks
    - [x] Linting (`black`, `flake8`, `pylint`)
    - [x] Security auditing (`safety`)
    - [x] License checking (`liccheck`)
    - [x] Unit testing (`unittests`)
    - [x] Integration testing (`pytest`, `pytest-coverage`)
    - [ ] Package management (`poetry`)
  - [x] Checks run against data processing scripts on every push and PR (`GitHub Actions`)

## Blockers/Questions

Blockers/questions will be checked off as they are addressed

- [x] I set up CI/CD on the existing code in `data-roadmap/` and also set up a fresh directory for data processing scripts in `data-pipeline/` per @switzersc-usds  suggestion. Is it okay to leave these as separate packages in the project, or do we want to combine them?
- [ ] `pre-commit` feels like the natural package to use for commit-based CI/CD checks. The one challenge is that when you run `pre-commit install` it installs the hook in the `.git` file at the root of this project, which messes up some of the references to configuration files in both `data-pipeline/` and `data-roadmap/`. I'm going to keep working through this, but it may be an inherent limitation of the mono-repo approach to this project.
- [x] Before adding a `data_processing_checks.yaml` to the `.github/workflows/` directory, I was hoping to work with you all to answer the first two questions. In the meantime I'll try to set the actions yaml file up to run CI/CD checks for both directories.
- [ ] I've saved the implementation of `poetry` for last because the [documentation for `liccheck2`](https://github.com/dhatim/python-license-check) specifies that it looks through `requirements.txt` to check for license compliance. Based on my current understanding of `poetry`, it removes the need for a `requirements.txt`, so adoption of that tool would break the usage of `liccheck` or require us to maintain a parallel `requirements.txt` in addition to the setup enforced by `poetry`. Would love the teams thoughts here, on how we wanted to balance that tradeoff
- [x] While the current `tox.ini` file runs the unit tests in `data-roadmap/` I did not reimplement them using `pytest` or apply coverage testing to them. Can this reimplementation be scoped into a different issue, or should it be addressed with this one?

## General Notes

- Please double check the `pyproject.toml` file for the configuration settings of `black`, `pylint`, and `liccheck` and the `.flake8` file for `flake8`. I used the recommended setting either from the library's documentation, or from articles that I linked to in the comments, but I'd love a second pair of eyes on those choices.
- I removed the direct reference to `requirements.txt` in `data-roadmap/setup.py` per the best practices outlined in this article from the [Python Packaging Authority](https://packaging.python.org/discussions/install-requires-vs-requirements/) but I recognize that may have been an intentional choice and can switch that back as needed.
- I also made some formatting adjustments to the python scripts in `data-roadmap/` to address some of the warnings thrown by `pylint` but disabled the remaining warnings within the two modules so that `tox` would pass

## Instructions to Review

1. Checkout this PR locally
2. Follow the installation instructions below for both `data-pipeline/` and `data-roadmap/` **Note** this will require setting up two virtual environments, one for each directory.
    1. Change directory into the package you want to test, i.e. `cd data-pipeline` or `cd data-roadmap`
    1. Create a new virtual environment by running `python -m venv env` in the command line.
    1. Activate the virtual environment by running `source env/bin/activate` (on Mac/Linux) or `.\env\Scripts\activate` (on Windows).
    1. Install this package in editable mode by running `pip install -e .` which makes changes made to scripts within this package available without re-installing it.
    1. Install the other dependencies required to contribute to the project by running `pip -r requirements.txt`
3. After following the installation instructions for each directory, run `tox` within the directory, all tests should pass. **Note** this needs to be run from the root of each directory, with that directory's virtual environment activated.

### CODE REVIEW

**Comment by widal001** [comment]
@NatHillardUSDS @switzersc-usds @esfoobar-usds I'm still working on a couple of the tasks outlined in Definition of Done above, but I'd love to get your input on a few of the blockers/questions I've listed in the Blockers/Questions section. 

I'll continue to push updates to this PR as I work through those blockers.

**Comment by widal001** [comment]
Screen Shot 2021-06-18 at 2 09 48 PM

I was able to get the GitHub actions for `tox` stood up and all checks pass on push, so I checked off that blocker. Unfortunately I'm still stuck on the other items mentioned in the the Blockers and Questions section. I'll mark this as "Ready to Merge" and perhaps we can evaluate which of those items we might want to address or descope.

**Comment by widal001** [comment]
Additionally, I've been working on migrating these packages from `setuptools` to `poetry` on [this branch](https://github.com/widal001/justice40-tool/tree/issue-57a-poetry-setup) and running into two major issues with both `liccheck` and `safety`: 

1. `liccheck` requires a `requirements.txt` file to be in the project directory otherwise it throws an error.
1. Despite forcing an update of `pip` by running `poetry run pip install -U pip` the version installed in the `tox` virtual environment when `poetry run safety check` is executed fails the safety check (see screenshot below)

I know that `poetry` is supposed to make package management easier, but it seems to have some compatibility issues with existing CI/CD tools that have been developed around the use of `requirements.txt` and `pip` How crucial is the adoption of `poetry` to this project? And are we willing to forgo or alter our use of `liccheck` and `safety` to make adopting `poetry` easier?

Screen Shot 2021-06-18 at 4 18 04 PM

**Comment by switzersc-usds** [comment]
Thanks so much for this PR, @widal001! This looks awesome, and I really appreciate how much documentation and thoughtful communication you've put into this. 

We've been pretty heads down this week setting up some of the data infrastructure, so haven't been able to take a thorough look yet. My goal is to check this out on Monday in advance of the open source community chat and move the related issue into our sprint backlog (which might involve some updates based on learnings from our data pipeline work this past sprint). Then we can address the open questions and blockers either during our community chat or find a followup time later in the week. 

Thanks again for the contribution, and happy Juneteenth!

**Comment by NatHillardUSDS** [comment]
Everything Shelby said, and! : just a side note that I've gotten around the need for a `requirements.txt` in the past with `poetry export --no-dev --without-hashes -f requirements.txt` which can bridge the gap without too much fuss, though it is yet another command to remember to run. I used this as a requirement for a `build` makefile command [here](https://github.com/department-of-veterans-affairs/Virtual-Regional-Office/blob/develop/hypertension/Makefile) as an example of a context (AWS SAM CLI) that would otherwise require a requirements.txt

File: data-pipeline/README.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     test_hello.py
+     test_your_new_module_1.py
+     your_new_module_2/
+        test_your_new_module_2_1.py
+        test_your_new_module_2_2.py
+
+   ```
+   > **Notes**
+   > - CI/CD checks will only pass if more than 90% of the code base is executed by the tests
+   > - Pytest requires the following naming conventions for [test discovery](https://docs.pytest.org/en/reorganize-docs/new-docs/user/naming_conventions.html)
```
Comment (ginabeena) [inline data-pipeline/README.md:89]: Yeah this is great, we should link to this from the main readme for getting started.

File: data-pipeline/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+pre-commit==2.13.0
+black==21.5b2
+flake8==3.9.2
+pylint==2.8.3
+tox==3.23.1
+pytest==6.2.4
+safety==1.10.3
+liccheck==0.6.0
+pytest-cov==2.12.1
```
Comment (ginabeena) [inline data-pipeline/requirements.txt:9]: Out of curiosity, what part of the toolchain consumes this file?

File: data-pipeline/tests/test_hello.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+from data_pipeline.hello import hello
+
+
+@pytest.mark.parametrize(
+    "name, expected",
+    [
+        ("World", "Hello, World!"),
+        ("Joe", "Hello, Joe!"),
+        ("Kamala", "Hello, Kamala!"),
```
Comment (ginabeena) [inline data-pipeline/tests/test_hello.py:11]: 😂

**Review by ginabeena — APPROVED** [review]

**Comment by esfoobar-usds** [comment]
@widal001 do you mind merging `main` onto this branch and re-review the PR? Would like your input of what's still missing.

**Comment by widal001** [comment]
@esfoobar-usds Just merged `main` into my feature branch on this. It should be good to merge since most of the CI/CD checks are on the `data-roadmap` directory and the `data-pipeline` directory. 

Even though it looks like you all have built out the `score/` directory with the data processing scripts, I'd recommend keeping `data-pipeline` in this PR to serve as a reference point for implementation of CI/CD checks on the `score/` directory. 

I'd be happy to work on re-implementing these checks for the `score/` directory as a part of issue #296 which I see as building on the work that you did in PR #260

**Comment by esfoobar-usds** [comment]
@widal001 it looks like there's a bunch of conflicts, can you review and let me know?

**Comment by widal001** [comment]
@esfoobar-usds I'm actually going to close this PR, since #405 tackles a piece of it, and the work I'm planning to do on #308 will capture the linting component of this. I'll try to add `safety` and `liccheck` to that as well so I can address #57 with my PR for 308

**Pull request closed without merging**